### PR TITLE
Initial version of Lit adapter for Tanstack form

### PR DESCRIPTION
**Pull request by Christian24**
> Builds on top of #588, please merge that first

This closes #498.

This is the first cut. 

It has tests, a basic example and the base-lib, there's also Material Web Components example. I added a little QuickStart guide. 

Open points:
* I am not 100% sure this is entirely type safe and I am not sure why it is not. It seems the field options passed `.field()` can be pretty much anything. Upon further investigation, this seems to be an issue with using an array, so maybe `DeepKeys` or its use is broken? @crutchcorn?

### CODE REVIEW

File: docs/framework/lit/quick-start.md
Comment (crutchcorn) [inline docs/framework/lit/quick-start.md:90]: FWIW (and I'd love to hear @fulopkovacs and @aadito123's opinion on this) I don't actually like this `bind` API very much.

I know this came up from convos with the Lit team, but I think our verbosity is the right move. If someone wants to write their own `bind` then that's fine (and we should document how to do so, copy+paste style), but we should not ship our own first-class adapter for any framework in this style IMO

**Review by fulopkovacs — COMMENTED** [review]
I haven't finished reviewing the PR, but I'm really happy that it exists! ☺️ 

According the the [awesome-lit](https://github.com/web-padawan/awesome-lit) list, the `Lit` ecosystem doesn't have a lot libraries for forms (I saw only one, but it was last published 2 years ago and has 55 weekly downloads on npm: [`@vaading/form`](https://www.npmjs.com/package/@vaadin/form)), so I think the community will find TanStack/form very useful. 

Thank you for your work! ☺️

File: docs/framework/lit/quick-start.md
Comment (aadito123) [reply docs/framework/lit/quick-start.md:90]: I have to agree here. There's a bit of a conflict between the TanStack way and the Lit way. For the sake of consistency across the other form adapters, we probably should not ship bind as part of the library.

Comment (Christian24) [reply docs/framework/lit/quick-start.md:90]: I am not sure I agree here. Isn't the idea of an adapter to adapt to a given ecosystem?

Comment (fulopkovacs) [reply docs/framework/lit/quick-start.md:90]: I'm not familiar with the Lit ecosystem and its conventions, but from what I get, all the other adapters aim to be thin wrappers around the core form lib and do not have API-s similar to `bind`. 

<details>
<summary>Handling the field's events in other adapters</summary>

- [React](https://github.com/TanStack/form/blob/eb4801c5202cbe3007401047f3c776351b0631b4/docs/framework/react/quick-start.md?plain=1#L35-L45)
```tsx
            <form.Field
              name="fullName"
              children={(field) => (
                <input
                  name={field.name}
                  value={field.state.value}
                  onBlur={field.handleBlur}
                  onChange={(e) => field.handleChange(e.target.value)}
                />
              )}
            />
```

- [Vue](https://github.com/TanStack/form/blob/eb4801c5202cbe3007401047f3c776351b0631b4/docs/framework/vue/quick-start.md?plain=1#L35-L44)
```vue
      <form.Field name="fullName">
        <template v-slot="{ field }">
          <input
            :name="field.name"
            :value="field.state.value"
            :onBlur="field.handleBlur"
            :onChange="(e) => field.handleChange(e.target.value)"
          />
        </template>
      </form.Field>
```

- [Solid](https://github.com/TanStack/form/blob/eb4801c5202cbe3007401047f3c776351b0631b4/docs/framework/solid/quick-start.md?plain=1#L34-L44)
```tsx
            <form.Field
              name="fullName"
              children={(field) => (
                <input
                  name={field().name}
                  value={field().state.value}
                  onBlur={field().handleBlur}
                  onInput={(e) => field().handleChange(e.target.value)}
                />
              )}
            />
```
</details>

Personally, I like the idea of having utilities like the `bind` directive to reduce the level of verbosity as a user, but they scare me a bit as a maintainer. 😆 The "thin wrapper"-philosophy helps us to not spread ourselves too thin, and allows us to work/iterate faster on the core feature set. I think this is pretty important, especially in the beginning. 

That said, it's also true that the implementation for the `bind` directive does not exactly have the size of a copy-pasteable code snippet: the `packages/lit-form/src/bind.ts` file has 139 lines. Maybe it could be in a separate package, outside of `tanstack/form`? (Or is it too small for that? 🤣)

**Review by fulopkovacs — COMMENTED** [review]
Ok, I'm done with reviewing the whole PR. I have never used Lit, but I tested the examples and read some parts of the Lit docs to get a better understanding on how Lit handles component states ([`ReactiveController` API](https://lit.dev/docs/api/controllers/)).

Based on my limited knowledge, this PR is a good first implementation of a Lit adapter for TanStack form. I added a few comments, but they are small ones. ☺️ 

That said, I should not approve the PR until the conversations started by @crutchcorn  on [the `bind` directive](https://github.com/TanStack/form/pull/577#discussion_r1485618134) and [the Lit adapter for `tanstack/store`](https://github.com/TanStack/form/pull/577#issuecomment-1937849056) have been resolved.

File: examples/lit/simple/src/index.ts
Comment (Christian24) [reply examples/lit/simple/src/index.ts:67]: You mean just add validations to the example?

File: packages/lit-form/src/control-value-accessor.ts
Comment (Christian24) [reply packages/lit-form/src/control-value-accessor.ts:46]: It is just me being stupid I think. I should add a `return` in the if clause ;)

File: packages/lit-form/src/control-value-accessor.ts
Comment (Christian24) [reply packages/lit-form/src/control-value-accessor.ts:31]: Sure. It is not part of the the demo so I did not think of it.

File: docs/framework/lit/quick-start.md
Comment (Christian24) [reply docs/framework/lit/quick-start.md:90]: I don't think putting it in another package is really going to be a solution...

File: examples/lit/simple/src/index.ts
Comment (Christian24) [reply examples/lit/simple/src/index.ts:68]: I have to revise my comment. Why do you want to display the error message as a `<p>`-tag here? The browser will automatically show it due to the `ControlValueAccessor` calling `setCustomValidity` and `reportValidity` on the element. See https://developer.mozilla.org/en-US/docs/Web/HTML/Constraint_validation for more info. Hence why I think having access to the element instance is so important. And hence why I built the `ControlValueAccessor`.

File: packages/lit-form/src/control-value-accessor.ts
Comment (Christian24) [reply packages/lit-form/src/control-value-accessor.ts]: Fixed. Good catch.

File: examples/lit/simple/src/index.ts
Comment (fulopkovacs) [reply examples/lit/simple/src/index.ts:68]: Ah, I see! You're right, we don't need to display the errors ourselves. Thanks for the explanation, now I think I understand this PR a bit better!


image

File: docs/framework/lit/quick-start.md
Comment (crutchcorn) [reply docs/framework/lit/quick-start.md:90]: I think this is the final thing we need to discuss for this PR - we're running into philosophical differences as maintainers (which is okay!).

I think we should avoid shipping `bind`-like APIs in order to avoid maintenance overhead on our end. As soon as we ship one instance of this for one adapter, it opens the potential floodgates for other adapters and then the requests for adapters in every UI library to come in.

I like the approach that @KevinVandy (please feel free to chime in here on this topic!) has done where he maintains different packages externally to TanStack Form for these adapters.

I strongly feel that this is the only way I'd feel comfortable merging this in 😅

Comment (crutchcorn) [reply docs/framework/lit/quick-start.md:90]: Resolving per a convo with @Christian24 that I am able to move forward with removing `bind` APIs for now

**Review by fulopkovacs — CHANGES_REQUESTED** [review]
There's only 1 thing that needs to be changed in `docs/config.json`, because the schema of the config.json files in the TanStack projects have been updated today. Other than that, it's all good.

**Comment by crutchcorn** [comment]
WTF I DID NO SUCH THING

**Comment by crutchcorn** [comment]
I did not close this PR or delete the upstream branch. Instead, I merged this PR:

https://github.com/TanStack/form/pull/588

Which this PR was supposed to be based off of.

I don't know what's happened - I've pinged GH (thru private channels). In the meantime, I have a backup somewhere, luckily

File: docs/config.json
Comment (crutchcorn) [reply docs/config.json]: This should be fixed now!

**Pull request merged**
